#📌 Extracción

In [54]:
import pandas as pd

telecom_data_url = "https://raw.githubusercontent.com/ingridcristh/challenge2-data-science-LATAM/main/TelecomX_Data.json"

telecom_dataframe = pd.read_json(telecom_data_url)

telecom_dataframe.head()

,customerID,Churn,customer,phone,internet,account
0,0002-ORFBO,No,"{'gender': 'Female', 'SeniorCitizen': 0, 'Part...","{'PhoneService': 'Yes', 'MultipleLines': 'No'}","{'InternetService': 'DSL', 'OnlineSecurity': '...","{'Contract': 'One year', 'PaperlessBilling': '..."
1,0003-MKNFE,No,"{'gender': 'Male', 'SeniorCitizen': 0, 'Partne...","{'PhoneService': 'Yes', 'MultipleLines': 'Yes'}","{'InternetService': 'DSL', 'OnlineSecurity': '...","{'Contract': 'Month-to-month', 'PaperlessBilli..."
2,0004-TLHLJ,Yes,"{'gender': 'Male', 'SeniorCitizen': 0, 'Partne...","{'PhoneService': 'Yes', 'MultipleLines': 'No'}","{'InternetService': 'Fiber optic', 'OnlineSecu...","{'Contract': 'Month-to-month', 'PaperlessBilli..."
3,0011-IGKFF,Yes,"{'gender': 'Male', 'SeniorCitizen': 1, 'Partne...","{'PhoneService': 'Yes', 'MultipleLines': 'No'}","{'InternetService': 'Fiber optic', 'OnlineSecu...","{'Contract': 'Month-to-month', 'PaperlessBilli..."
4,0013-EXCHZ,Yes,"{'gender': 'Female', 'SeniorCitizen': 1, 'Part...","{'PhoneService': 'Yes', 'MultipleLines': 'No'}","{'InternetService': 'Fiber optic', 'OnlineSecu...","{'Contract': 'Month-to-month', 'PaperlessBilli..."


#🔧 Transformación

In [55]:
customer_data = telecom_dataframe["customer"].apply(pd.Series)
phone_data = telecom_dataframe["phone"].apply(pd.Series)
internet_data = telecom_dataframe["internet"].apply(pd.Series)
account_data = telecom_dataframe["account"].apply(pd.Series)

telecom_dataframe = pd.concat(
    [
        telecom_dataframe.drop(columns=["customer", "phone", "internet", "account"]),
        customer_data,
        phone_data,
        internet_data,
        account_data
    ],
    axis=1
)

In [56]:
telecom_dataframe = telecom_dataframe[telecom_dataframe["Churn"] != ""]

In [57]:
telecom_dataframe["Charges"].head()

,Charges
0,"{'Monthly': 65.6, 'Total': '593.3'}"
1,"{'Monthly': 59.9, 'Total': '542.4'}"
2,"{'Monthly': 73.9, 'Total': '280.85'}"
3,"{'Monthly': 98.0, 'Total': '1237.85'}"
4,"{'Monthly': 83.9, 'Total': '267.4'}"


In [58]:
charges_data = telecom_dataframe["Charges"].apply(pd.Series)

telecom_dataframe["MonthlyCharges"] = charges_data["Monthly"]
telecom_dataframe["TotalCharges"] = pd.to_numeric(charges_data["Total"], errors="coerce")

telecom_dataframe = telecom_dataframe.drop(columns=["Charges"])

In [59]:
telecom_dataframe.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7043 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   Churn             7043 non-null   object 
 2   gender            7043 non-null   object 
 3   SeniorCitizen     7043 non-null   int64  
 4   Partner           7043 non-null   object 
 5   Dependents        7043 non-null   object 
 6   tenure            7043 non-null   int64  
 7   PhoneService      7043 non-null   object 
 8   MultipleLines     7043 non-null   object 
 9   InternetService   7043 non-null   object 
 10  OnlineSecurity    7043 non-null   object 
 11  OnlineBackup      7043 non-null   object 
 12  DeviceProtection  7043 non-null   object 
 13  TechSupport       7043 non-null   object 
 14  StreamingTV       7043 non-null   object 
 15  StreamingMovies   7043 non-null   object 
 16  Contract          7043 non-null   object 
 17  

In [60]:
telecom_dataframe["TotalCharges"] = telecom_dataframe["TotalCharges"].fillna(0)

In [61]:
telecom_dataframe.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7043 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   Churn             7043 non-null   object 
 2   gender            7043 non-null   object 
 3   SeniorCitizen     7043 non-null   int64  
 4   Partner           7043 non-null   object 
 5   Dependents        7043 non-null   object 
 6   tenure            7043 non-null   int64  
 7   PhoneService      7043 non-null   object 
 8   MultipleLines     7043 non-null   object 
 9   InternetService   7043 non-null   object 
 10  OnlineSecurity    7043 non-null   object 
 11  OnlineBackup      7043 non-null   object 
 12  DeviceProtection  7043 non-null   object 
 13  TechSupport       7043 non-null   object 
 14  StreamingTV       7043 non-null   object 
 15  StreamingMovies   7043 non-null   object 
 16  Contract          7043 non-null   object 
 17  

#📊 Carga y análisis

In [62]:
telecom_dataframe["Churn"].value_counts()

,count
Churn,
No,5174
Yes,1869


In [63]:
telecom_dataframe["Churn"].value_counts(normalize=True)

,proportion
Churn,
No,0.73463
Yes,0.26537


El análisis inicial muestra que aproximadamente 26.5 % de los clientes han abandonado el servicio, mientras que 73.5 % permanecen activos.
Esto indica que cerca de 1 de cada 4 clientes termina cancelando su servicio, lo cual representa una tasa de abandono considerable para una empresa de telecomunicaciones



In [64]:
pd.crosstab(
    telecom_dataframe["Contract"],
    telecom_dataframe["Churn"],
    normalize="index"
)

Churn,No,Yes
Contract,,
Month-to-month,0.572903,0.427097
One year,0.887305,0.112695
Two year,0.971681,0.028319


Relación entre tipo de contrato y abandono

El análisis muestra que el tipo de contrato tiene una fuerte relación con la tasa de abandono de clientes. Los clientes con contratos mes a mes presentan la mayor probabilidad de cancelar el servicio, con aproximadamente 42.7 % de abandono.

En contraste, los clientes con contratos de un año presentan una tasa de abandono cercana al 11 %, mientras que aquellos con contratos de dos años muestran una tasa significativamente menor, cercana al 3 %.

Esto sugiere que los contratos de mayor duración actúan como un mecanismo importante de retención de clientes, reduciendo considerablemente la probabilidad de abandono.

In [65]:
telecom_dataframe.groupby("Churn")["tenure"].mean()

,tenure
Churn,
No,37.569965
Yes,17.979133


El análisis muestra que los clientes que abandonan el servicio presentan una antigüedad promedio significativamente menor que aquellos que permanecen en la compañía.

Esto indica que los primeros meses del servicio representan un periodo crítico para la retención de clientes, ya que los clientes con mayor tiempo utilizando el servicio tienden a permanecer más tiempo con la empresa.

#📄Informe final

El análisis exploratorio del conjunto de datos permitió identificar algunos factores relacionados con el abandono de clientes en el servicio de telecomunicaciones.

En primer lugar, se observó que aproximadamente 26.5 % de los clientes han cancelado el servicio, lo que representa una proporción considerable dentro de la base de usuarios.

Asimismo, el análisis mostró que el tipo de contrato tiene una fuerte relación con el abandono, ya que los clientes con contratos mes a mes presentan tasas de cancelación significativamente mayores en comparación con aquellos que poseen contratos de uno o dos años.

Finalmente, se identificó que la antigüedad del cliente también influye en la probabilidad de abandono, ya que los clientes con menor tiempo utilizando el servicio tienden a cancelar con mayor frecuencia.

En conjunto, estos resultados sugieren que estrategias orientadas a fomentar contratos de mayor duración y mejorar la experiencia del cliente durante los primeros meses de servicio podrían contribuir a reducir la tasa de abandono.